dlls/gdi32/objects.c

```diff
@@ -971,10 +971,78 @@ NTSTATUS WINAPI D3DKMTOpenAdapterFromGdiDisplayName( D3DKMT_OPENADAPTERFROMGDIDI
     return status;
 }
 
-NTSTATUS WINAPI D3DKMTEnumAdapters2( const void *param )
+NTSTATUS WINAPI D3DKMTEnumAdapters2( D3DKMT_ENUMADAPTERS2 *enumAdapters )
 {
-    FIXME( "param %p stub.\n", param );
-    return STATUS_NOT_SUPPORTED;
+    NTSTATUS status = STATUS_SUCCESS;
+    SP_DEVINFO_DATA device_data;
+    DEVPROPTYPE type;
+    HDEVINFO devinfo;
+    UINT dev_count = 0;
+    HANDLE mutex;
+
+    TRACE("(%p)\n", enumAdapters);
+
+    mutex = get_display_device_init_mutex();
+    devinfo = SetupDiGetClassDevsW(&GUID_DEVCLASS_DISPLAY, L"PCI", NULL, 0);
```

Comment:
Does the `"PCI"` enumerator also include software devices like the Microsoft Basic Display Adapter with the WARP software Direct3D UMD, as well as non-PCI GPUs such as the Host1x-attached Nvidia Tegra 3's GPU in the Microsoft Surface RT?

For enumerating all display devices, `CM_Get_Device_Interface_List` for `GUID_DISPLAY_DEVICE_ARRIVAL` can be used alternatively, it does include the software adapter.

Reply:
According to https://learn.microsoft.com/en-us/windows-hardware/drivers/install/enumerating-installed-device-interface-classes, `SetupDiGetClassDevs` also can be used for the purposes of enumeration, but the `Enumerator` argument can be `NULL`. And yeah, since you can query the LUID this way (via `DEVPROPKEY_GPU_LUID`) without actually calling `D3DKMTOpenAdapterFromDeviceName`, it should be more straightforward than using the Configuration Manager. Is the `"PCI"` enumerator needed for anything important in Wine, and what happens if you don't pass it?